File: java-project-euler-100/src/main/java/net/talaatharb/projecteuler100/P015.java
package net.talaatharb.projecteuler100;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

/**
 * P015
 * 
 * @author mharb
 *
 */
public class P015 extends Solution {

	/**
	 * Problem solution entry point
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		new P015().run();
	}

	private final Map<Integer, BigInteger> memoryCache;

	/**
	 * Public constructor
	 */
	public P015() {
		memoryCache = new HashMap<>();
		memoryCache.put(0, BigInteger.valueOf(1));
	}

	private BigInteger factorial(final int n) {
		if (memoryCache.containsKey(n)) {
			return memoryCache.get(n);
		}
		final BigInteger data = BigInteger.valueOf(n).multiply(factorial(n - 1));
		memoryCache.put(n, data);
		return data;
	}

	@Override
	public Number solve() {
		problemNumber = 15;
		final int n = 20;

		for (int i = 1; i < (2 * n); i++) {
			factorial(i);
		}

		return factorial(2 * n).divide((factorial(n).multiply(factorial(n))));
	}

}

File: typescript-project-euler-100/src/p015.ts
import { Solution } from "./Solution";

export class P015 extends Solution {

    memoryCache: Map<number, number> = new Map<number, number>();

    constructor() {
        super();
        this.memoryCache.set(0, 1);
    }

    factorial(n: number) : number{
		if (this.memoryCache.has(n)) {
			return this.memoryCache.get(n) as number;
		}
		const data = n * (this.factorial(n - 1));
		this.memoryCache.set(n, data);
		return data;
	}


    solve() {
        this.problemNumber = 15;
        const n = 20;

		for (let i = 1; i < (2 * n); i++) {
			this.factorial(i);
		}

		return this.factorial(2 * n) / ((this.factorial(n) * (this.factorial(n))));
    }
}
